src/librustc_typeck/check/mod.rs

```diff
@@ -1709,6 +1717,10 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
     }
 
     pub fn expr_ty(&self, ex: &hir::Expr) -> Ty<'tcx> {
+        if let Some(&adjustment::AdjustNeverToAny(ref t))
+                = self.tables.borrow().adjustments.get(&ex.id) {
+            return t;
+        }
         match self.tables.borrow().node_types.get(&ex.id) {
```

Comment:
I am somewhat surprised that `node_ty` was not changed (and also surprised that `expr_ty` doesn't just call `node_ty`, I wonder when/why that changed).

@eddyb any thoughts on this point?